common/src/main/java/io/netty/util/internal/MpscLinkedQueue.java

```diff
@@ -248,53 +256,37 @@ public E element() {
         throw new NoSuchElementException();
     }
 
-    @Override
-    public Object[] toArray() {
-        final Object[] array = new Object[size()];
-        final Iterator<E> it = iterator();
-        for (int i = 0; i < array.length; i ++) {
-            if (it.hasNext()) {
-                array[i] = it.next();
-            } else {
-                return Arrays.copyOf(array, i);
+    private List<E> toList(int initialCapacity) {
+        List<E> elements = new ArrayList<E>(initialCapacity);
+
+        MpscLinkedQueueNode<E> n = peekNode();
+        for (;;) {
+            if (n == null) {
+                break;
             }
+            E value = n.value();
+            if (value == null) {
+                break;
+            }
+            elements.add(value);
+            MpscLinkedQueueNode<E> next = n.next();
+            if (n == next) {
+                break;
+            }
+            n = next;
         }
-        return array;
+        return elements;
+    }
+
+    @Override
+    public Object[] toArray() {
+        return toList(16).toArray();
```

Comment:
Where did `16` come from?  Is this a value we have explicitly decided is good...or are we just using it because it is also the default value of some common collection implementations in Java? If we have justification why this makes sense...then never mind me (but add a comment :) ).  Otherwise if we don't have any rational for the size and just choose it because that is what default implementations choose then lets just let the implementation choose the default (i.e. `new ArrayList<>()`) and change the parameter from a `int` to a `List<>`.